server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java

```diff
@@ -459,6 +459,17 @@ protected MapperBuilderContext createChildContext(MapperBuilderContext mapperBui
     }
 
     public ObjectMapper merge(Mapper mergeWith, MergeReason reason, MapperBuilderContext parentBuilderContext) {
+        if (mergeWith instanceof ObjectMapper == false) {
+            MapperErrors.throwObjectMappingConflictError(mergeWith.name());
+        }
+        if (this instanceof NestedObjectMapper == false && mergeWith instanceof NestedObjectMapper) {
+            // TODO stop NestedObjectMapper extending ObjectMapper?
+            MapperErrors.throwNestedMappingConflictError(mergeWith.name());
+        }
```

Comment:
Hopefully I was able to isolate these checks to the only place where they are really needed. We previously would perform them e.g. in RootObjectMapper, where we'd later cast the mapper to RootObjectMapper.